### PR TITLE
OJ-2736: `/events` endpoint

### PR DESCRIPTION
**Pull request by surajkumar**
### What changed
Created a new endpoint called `/events` which allows you to query the "audit events table".
There is also a new domain name associated with this endpoint in the format:  `<stackName>.<Environment>.account.gov.uk/` e.g. `https://test-resources.review-hc.dev.account.gov.uk/`

This events API takes 2 query string parameters; partitionKey and sortKey. 

### Why did it change
To support the test strategy with the test harness approach. 

Sample API output:
```
[{"sortKey":{"S":"TXMA#IPV_HMRC_RECORD_CHECK_CRI_START#1726084120#69de99e1-7a75-4abf-b355-7b9e3c3c52b9"},"event":{"S":"{\"component_id\":\"https://review-hc.dev.account.gov.uk\",\"event_name\":\"IPV_HMRC_RECORD_CHECK_CRI_START\",\"extensions\":{\"evidence\":{\"context\":\"identity_check\"}},\"restricted\":{\"device_information\":{\"encoded\":\"eyJhdWRpdF9oZWFkZXJfdmVyc2lvbiI6IjEuNC4yIiwicmVxdWVzdF90aW1lc3RhbXBfbXMiOjE3MjYwODQxMjA2NDcsImlwX2FkZHJlc3MiOiIxMy40Mi4yNy4yNTEiLCJjb3VudHJ5X2NvZGUiOiJHQiIsInVzZXJfYWdlbnQiOiJNb3ppbGxhLzUuMCAoWDExOyBMaW51eCB4ODZfNjQpIEFwcGxlV2ViS2l0LzUzNy4zNiAoS0hUTUwsIGxpa2UgR2Vja28pIEhlYWRsZXNzQ2hyb21lLzEyMS4wLjYxNjcuMTM5IFNhZmFyaS81MzcuMzYgQ2xvdWRXYXRjaFN5bnRoZXRpY3MvYXJuOmF3czpzeW50aGV0aWNzOmV1LXdlc3QtMjo1NjI2NzAyNjY0OTY6Y2FuYXJ5Omthcm9sLXN0LWhhcHB5IiwiamEzX2ZpbmdlcnByaW50IjoiOTUzYTVlM2JiMTc0ZDUxNjljNGVjNzJiYmM3MTc5YWIiLCJjb25uZWN0aW9uX3BvcnQiOjU3NjkwfQ\"}},\"timestamp\":1726084120,\"event_timestamp_ms\":1726084120820,\"user\":{\"govuk_signin_journey_id\":\"64c8253a-959c-4e3d-b882-5418d0db7f92\",\"ip_address\":\"13.42.27.251, 10.0.12.206\",\"persistent_session_id\":\"b80f685e-a655-4aba-aa51-571bdbe4be49\",\"session_id\":\"0831772a-7ac2-42a5-8aca-3aca4ffcb8f8\",\"user_id\":\"urn:fdc:gov.uk:2022:b2f0924a-6e4d-44bc-84fe-652c660c16a3\"}}"},"partitionKey":{"S":"SESSION#0831772a-7ac2-42a5-8aca-3aca4ffcb8f8"}}]
```

### Issue tracking
- [OJ-2736](https://govukverify.atlassian.net/browse/OJ-2736)

[OJ-2736]: https://govukverify.atlassian.net/browse/OJ-2736?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: test-resources/test-harness/infrastructure/template.yaml [hunk 1]
```diff
@@ -135,4 +135,70 @@ Resources:
...
       TimeToLiveSpecification:
         AttributeName: expiryDate
-        Enabled: true
\ No newline at end of file
+        Enabled: true
+
+  APIGateway:
+    Type: AWS::Serverless::Api
+    Properties:
+      Name: !Sub ${AWS::StackName}-apigw
```
Comment (jkunle) [inline test-resources/test-harness/infrastructure/template.yaml:143]: We shouldn't give explicit names unless we absolutely need to. It is go practice to all SAM to do this

File: test-resources/test-harness/infrastructure/public-api.yaml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                  "KeyConditionExpression": "partitionKey = :partitionKey AND sortKey = :sortKey",
+                  "ExpressionAttributeValues": {
+                    ":partitionKey": {
+                      "S": "$partitionKey"
+                    },
+                    ":sortKey": {
+                      "S": "$sortKey"
+                    }
+                  }
+                }
```
Comment (jkunle) [inline test-resources/test-harness/infrastructure/public-api.yaml:59]: Nice, I know it is VTL but this is acceptable

**Review by Caitlin-cooling — DISMISSED** [review]
Nice! Like the updates to the samconfig too - thanks

File: test-resources/test-harness/infrastructure/public-api.yaml [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    - Effect: "Deny"
+      Principal:
+        AWS:  "*"
+      Action: "execute-api:Invoke"
+      Resource: "execute-api:/*"
+      Condition:
+        StringNotEquals:
+          "aws:PrincipalAccount":
+            - "${AWS::AccountId}"
+    - Effect: "Allow"
```
[the remark this answers is not in the document]
Comment (surajkumar) [reply test-resources/test-harness/infrastructure/public-api.yaml:91]: Good spot, this was a c&p from one of the other teams and didn't notice the conflicting permission

**Review by jkunle — APPROVED** [review]

File: test-resources/test-harness/infrastructure/template.yaml [hunk 4]
```diff
@@ -20,6 +20,26 @@ Parameters:
     Description: "The stack containing the TXMA infrastructure"
     Type: String
     Default: txma-infrastructure
+  Environment:
+    Type: String
+    Default: dev
+    AllowedValues: [dev, localdev, build, staging]
```
Comment (Caitlin-cooling) [inline test-resources/test-harness/infrastructure/template.yaml:26]: Shouldn't be allowed in staging as we're integrated with TxMA there

File: test-resources/test-harness/infrastructure/template.yaml [hunk 5]
```diff
@@ -135,4 +168,132 @@ Resources:
...
+          PolicyDocument:
+            Version: 2012-10-17
+            Statement:
+              - Effect: Allow
+                Action:
+                  - dynamodb:Query
+                Resource: !GetAtt AuditEventsTable.Arn
+
+Outputs:
+  TestHarnessExecuteUrl:
```
Comment (Caitlin-cooling) [inline test-resources/test-harness/infrastructure/template.yaml:293]: Lovely!

File: test-resources/test-harness/infrastructure/template.yaml (hunk 4 above)
Comment (Caitlin-cooling) [reply test-resources/test-harness/infrastructure/template.yaml:26]: Actually I guess if we start using it for other things (eg DDB) maybe we would want in staging will resolve this comment

**Pull request merged**